File: clients/go/client_test.go
package aichat_test

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"reflect"
	"testing"
	"time"

	aichat "github.com/contextgrip-io/agent-sdk/clients/go"
)

func mustTime(t *testing.T, s string) time.Time {
	t.Helper()
	ts, err := time.Parse(time.RFC3339, s)
	if err != nil {
		t.Fatalf("parse time %q: %v", s, err)
	}
	return ts
}

func TestStatus(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.URL.Path != "/api/v1/status" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `{"version":"0.1.0","model":"claude-sonnet-4-5","engine":"postgresql","ready":true,"features":["chat","agent","board"],"writesEnabled":true}`)
	}))
	defer srv.Close()

	got, err := aichat.New(srv.URL, "tok").Status(context.Background())
	if err != nil {
		t.Fatalf("Status: %v", err)
	}
	want := aichat.Status{
		Version:       "0.1.0",
		Model:         "claude-sonnet-4-5",
		Engine:        "postgresql",
		Ready:         true,
		Features:      []string{"chat", "agent", "board"},
		WritesEnabled: true,
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("Status = %+v, want %+v", got, want)
	}
}

func TestAsk(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			t.Errorf("method = %s, want POST", r.Method)
		}
		if r.URL.Path != "/api/v1/ask" {
			t.Errorf("path = %s, want /api/v1/ask", r.URL.Path)
		}
		if got := r.Header.Get("Authorization"); got != "Bearer test-token" {
			t.Errorf("Authorization = %q, want %q", got, "Bearer test-token")
		}
		if got := r.Header.Get("Content-Type"); got != "application/json" {
			t.Errorf("Content-Type = %q, want application/json", got)
		}
		var req aichat.AskRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Errorf("decode request: %v", err)
		}
		want := aichat.AskRequest{Question: "how many users signed up last week?", ConversationID: "c1"}
		if req != want {
			t.Errorf("request = %+v, want %+v", req, want)
		}
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `{
			"conversationId": "c1",
			"userMessageId": "u1",
			"assistantMessageId": "a1",
			"sql": "SELECT count(*) FROM users WHERE created_at > now() - interval '7 days'",
			"result": {"columns":["count"],"rowSample":[[42]],"rowCount":1,"truncated":false,"executionTimeMs":12},
			"answer": "42 users signed up last week."
		}`)
	}))
	defer srv.Close()

	c := aichat.New(srv.URL, "test-token")
	got, err := c.Ask(context.Background(), aichat.AskRequest{
		Question:       "how many users signed up last week?",
		ConversationID: "c1",
	})
	if err != nil {
		t.Fatalf("Ask: %v", err)
	}
	want := aichat.AskResponse{
		ConversationID:     "c1",
		UserMessageID:      "u1",
		AssistantMessageID: "a1",
		SQL:                "SELECT count(*) FROM users WHERE created_at > now() - interval '7 days'",
		Result: &aichat.ResultSummary{
			Columns:         []string{"count"},
			RowSample:       [][]any{{float64(42)}},
			RowCount:        1,
			Truncated:       false,
			ExecutionTimeMs: 12,
		},
		Answer: "42 users signed up last week.",
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("AskResponse = %+v, want %+v", got, want)
	}
}

func TestAskExecutionFailure(t *testing.T) {
	// Failed query execution is NOT an HTTP error: 200 with resultError set.
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `{
			"conversationId": "c1",
			"userMessageId": "u1",
			"assistantMessageId": "a1",
			"sql": "SELECT nope FROM users",
			"resultError": "column \"nope\" does not exist",
			"answer": "The query failed: the column does not exist."
		}`)
	}))
	defer srv.Close()

	got, err := aichat.New(srv.URL, "tok").Ask(context.Background(), aichat.AskRequest{Question: "q"})
	if err != nil {
		t.Fatalf("Ask: %v", err)
	}
	if got.Result != nil {
		t.Errorf("Result = %+v, want nil", got.Result)
	}
	if want := `column "nope" does not exist`; got.ResultError != want {
		t.Errorf("ResultError = %q, want %q", got.ResultError, want)
	}
}

func TestAPIErrorMapping(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusUnauthorized)
		io.WriteString(w, `{"error":"missing or invalid bearer token","code":"UNAUTHORIZED"}`)
	}))
	defer srv.Close()

	_, err := aichat.New(srv.URL, "bad-token").Status(context.Background())
	if err == nil {
		t.Fatal("Status: expected error, got nil")
	}
	var apiErr *aichat.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("errors.As(*APIError) failed for %T: %v", err, err)
	}
	if apiErr.StatusCode != http.StatusUnauthorized {
		t.Errorf("StatusCode = %d, want 401", apiErr.StatusCode)
	}
	if apiErr.Code != "UNAUTHORIZED" {
		t.Errorf("Code = %q, want UNAUTHORIZED", apiErr.Code)
	}
	if apiErr.Message != "missing or invalid bearer token" {
		t.Errorf("Message = %q", apiErr.Message)
	}
	if apiErr.Error() == "" {
		t.Error("Error() returned empty string")
	}
}

func TestAPIErrorNonJSONBody(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadGateway)
		io.WriteString(w, "upstream exploded")
	}))
	defer srv.Close()

	_, err := aichat.New(srv.URL, "tok").Status(context.Background())
	var apiErr *aichat.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("errors.As(*APIError) failed for %T: %v", err, err)
	}
	if apiErr.StatusCode != http.StatusBadGateway || apiErr.Code != "" || apiErr.Message != "upstream exploded" {
		t.Errorf("APIError = %+v", apiErr)
	}
}

func TestConversationsCRUD(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("GET /api/v1/conversations", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `[
			{"id":"c2","title":"Newest","createdAt":"2026-07-15T12:00:00Z","updatedAt":"2026-07-15T13:00:00Z"},
			{"id":"c1","title":"Older","createdAt":"2026-07-14T09:00:00Z","updatedAt":"2026-07-14T09:05:00Z"}
		]`)
	})
	mux.HandleFunc("GET /api/v1/conversations/{id}", func(w http.ResponseWriter, r *http.Request) {
		if r.PathValue("id") != "c1" {
			w.WriteHeader(http.StatusNotFound)
			io.WriteString(w, `{"error":"conversation not found","code":"NOT_FOUND"}`)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `{
			"conversation": {"id":"c1","title":"Older","createdAt":"2026-07-14T09:00:00Z","updatedAt":"2026-07-14T09:05:00Z"},
			"messages": [
				{"id":"u1","role":"user","text":"how many users?","createdAt":"2026-07-14T09:00:00Z"},
				{"id":"a1","role":"assistant","text":"There are 42 users.","sql":"SELECT count(*) FROM users",
				 "result":{"columns":["count"],"rowSample":[[42]],"rowCount":1,"truncated":false,"executionTimeMs":9},
				 "createdAt":"2026-07-14T09:00:05Z"}
			]
		}`)
	})
	var deleted string
	mux.HandleFunc("DELETE /api/v1/conversations/{id}", func(w http.ResponseWriter, r *http.Request) {
		deleted = r.PathValue("id")
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `{"deleted":true}`)
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	c := aichat.New(srv.URL, "tok")
	ctx := context.Background()

	list, err := c.ListConversations(ctx)
	if err != nil {
		t.Fatalf("ListConversations: %v", err)
	}
	if len(list) != 2 || list[0].ID != "c2" || list[1].ID != "c1" {
		t.Errorf("ListConversations = %+v", list)
	}
	if !list[0].UpdatedAt.Equal(mustTime(t, "2026-07-15T13:00:00Z")) {
		t.Errorf("UpdatedAt = %v", list[0].UpdatedAt)
	}

	detail, err := c.GetConversation(ctx, "c1")
	if err != nil {
		t.Fatalf("GetConversation: %v", err)
	}
	if detail.Conversation.ID != "c1" || len(detail.Messages) != 2 {
		t.Fatalf("GetConversation = %+v", detail)
	}
	assistant := detail.Messages[1]
	if assistant.Role != "assistant" || assistant.SQL != "SELECT count(*) FROM users" {
		t.Errorf("assistant message = %+v", assistant)
	}
	if assistant.Result == nil || assistant.Result.RowCount != 1 || assistant.Result.ExecutionTimeMs != 9 {
		t.Errorf("assistant result = %+v", assistant.Result)
	}

	var apiErr *aichat.APIError
	if _, err := c.GetConversation(ctx, "missing"); !errors.As(err, &apiErr) || apiErr.Code != "NOT_FOUND" {
		t.Errorf("GetConversation(missing) err = %v, want NOT_FOUND APIError", err)
	}

	if err := c.DeleteConversation(ctx, "c1"); err != nil {
		t.Fatalf("DeleteConversation: %v", err)
	}
	if deleted != "c1" {
		t.Errorf("deleted id = %q, want c1", deleted)
	}
}

func TestTokenAdmin(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("GET /api/v1/tokens", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `[
			{"id":"t1","label":"ci","fingerprint":"deadbeef","createdAt":"2026-07-01T00:00:00Z","lastUsedAt":"2026-07-15T08:00:00Z"},
			{"id":"t2","label":"fresh","fingerprint":"cafef00d","createdAt":"2026-07-10T00:00:00Z"}
		]`)
	})
	mux.HandleFunc("POST /api/v1/tokens", func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Content-Type"); got != "application/json" {
			t.Errorf("Content-Type = %q, want application/json", got)
		}
		var body struct {
			Label string `json:"label"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.Label != "reporting" {
			t.Errorf("body = %+v, err = %v, want label reporting", body, err)
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		io.WriteString(w, `{"id":"t3","label":"reporting","fingerprint":"01234567","createdAt":"2026-07-15T12:00:00Z","token":"tok_raw_value_shown_once"}`)
	})
	var revoked string
	mux.HandleFunc("DELETE /api/v1/tokens/{id}", func(w http.ResponseWriter, r *http.Request) {
		revoked = r.PathValue("id")
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, `{"deleted":true}`)
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	c := aichat.New(srv.URL, "primary-token")
	ctx := context.Background()

	tokens, err := c.ListTokens(ctx)
	if err != nil {
		t.Fatalf("ListTokens: %v", err)
	}
	if len(tokens) != 2 {
		t.Fatalf("ListTokens returned %d tokens", len(tokens))
	}
	if tokens[0].LastUsedAt == nil || !tokens[0].LastUsedAt.Equal(mustTime(t, "2026-07-15T08:00:00Z")) {
		t.Errorf("tokens[0].LastUsedAt = %v", tokens[0].LastUsedAt)
	}
	if tokens[1].LastUsedAt != nil {
		t.Errorf("tokens[1].LastUsedAt = %v, want nil", tokens[1].LastUsedAt)
	}

	created, err := c.CreateToken(ctx, "reporting")
	if err != nil {
		t.Fatalf("CreateToken: %v", err)
	}
	if created.ID != "t3" || created.Label != "reporting" || created.Token != "tok_raw_value_shown_once" {
		t.Errorf("CreateToken = %+v", created)
	}

	if err := c.RevokeToken(ctx, "t1"); err != nil {
		t.Fatalf("RevokeToken: %v", err)
	}
	if revoked != "t1" {
		t.Errorf("revoked id = %q, want t1", revoked)
	}
}

func TestNoAuthorizationHeaderWhenTokenEmpty(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if _, ok := r.Header["Authorization"]; ok {
			t.Error("Authorization header sent for empty token")
		}
		io.WriteString(w, `{"version":"0.1.0","model":"m","engine":"postgresql","ready":true}`)
	}))
	defer srv.Close()

	if _, err := aichat.New(srv.URL, "").Status(context.Background()); err != nil {
		t.Fatalf("Status: %v", err)
	}
}

File: clients/typescript/test/client.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import { AiChatClient, AiChatError } from '../dist/index.js';
import { startStub, readBody, sendJson } from './stub.js';

test('ask() posts the question with auth + JSON headers and returns the answer', async () => {
  let seen;
  const stub = await startStub(async (req, res) => {
    seen = {
      method: req.method,
      url: req.url,
      authorization: req.headers.authorization,
      contentType: req.headers['content-type'],
      body: await readBody(req),
    };
    sendJson(res, 200, {
      conversationId: 'conv-1',
      userMessageId: 'msg-u1',
      assistantMessageId: 'msg-a1',
      sql: 'SELECT count(*) FROM orders',
      result: {
        columns: ['count'],
        rowSample: [[42]],
        rowCount: 1,
        truncated: false,
        executionTimeMs: 7,
      },
      answer: 'There are 42 orders.',
    });
  });
  try {
    const client = new AiChatClient({ baseUrl: stub.url, token: 'tok-abc' });
    const resp = await client.ask({ question: 'How many orders?' });

    assert.equal(seen.method, 'POST');
    assert.equal(seen.url, '/api/v1/ask');
    assert.equal(seen.authorization, 'Bearer tok-abc');
    assert.match(seen.contentType, /^application\/json/);
    assert.deepEqual(JSON.parse(seen.body), { question: 'How many orders?' });

    assert.equal(resp.conversationId, 'conv-1');
    assert.equal(resp.userMessageId, 'msg-u1');
    assert.equal(resp.assistantMessageId, 'msg-a1');
    assert.equal(resp.sql, 'SELECT count(*) FROM orders');
    assert.deepEqual(resp.result.rowSample, [[42]]);
    assert.equal(resp.answer, 'There are 42 orders.');
  } finally {
    await stub.close();
  }
});

test('ask() passes conversationId through and surfaces resultError on HTTP 200', async () => {
  let body;
  const stub = await startStub(async (req, res) => {
    body = JSON.parse(await readBody(req));
    sendJson(res, 200, {
      conversationId: 'conv-2',
      userMessageId: 'u',
      assistantMessageId: 'a',
      sql: 'SELECT * FROM missing_table',
      resultError: 'relation "missing_table" does not exist',
      answer: 'That table does not exist.',
    });
  });
  try {
    const client = new AiChatClient({ baseUrl: stub.url, token: 't' });
    const resp = await client.ask({ question: 'q', conversationId: 'conv-2' });
    assert.deepEqual(body, { question: 'q', conversationId: 'conv-2' });
    assert.equal(resp.result, undefined);
    assert.equal(resp.resultError, 'relation "missing_table" does not exist');
  } finally {
    await stub.close();
  }
});

test('401 maps to AiChatError with status and code UNAUTHORIZED', async () => {
  const stub = await startStub((req, res) => {
    sendJson(res, 401, { error: 'missing or invalid bearer token', code: 'UNAUTHORIZED' });
  });
  try {
    const client = new AiChatClient({ baseUrl: stub.url, token: 'bad' });
    await assert.rejects(client.status(), (err) => {
      assert.ok(err instanceof AiChatError);
      assert.equal(err.name, 'AiChatError');
      assert.equal(err.status, 401);
      assert.equal(err.code, 'UNAUTHORIZED');
      assert.equal(err.message, 'missing or invalid bearer token');
      return true;
    });
  } finally {
    await stub.close();
  }
});

test('non-JSON error bodies still produce an AiChatError with the status', async () => {
  const stub = await startStub((req, res) => {
    res.writeHead(502, { 'content-type': 'text/html' });
    res.end('<html>Bad Gateway</html>');
  });
  try {
    const client = new AiChatClient({ baseUrl: stub.url, token: 't' });
    await assert.rejects(client.status(), (err) => {
      assert.ok(err instanceof AiChatError);
      assert.equal(err.status, 502);
      assert.equal(err.code, undefined);
      return true;
    });
  } finally {
    await stub.close();
  }
});

test('status() hits GET /api/v1/status; no auth header without a token', async () => {
  let seen;
  const stub = await startStub((req, res) => {
    seen = { method: req.method, url: req.url, authorization: req.headers.authorization };
    sendJson(res, 200, {
      version: '0.1.0',
      model: 'claude-opus-4-8',
      engine: 'postgresql',
      ready: true,
      features: ['chat', 'agent', 'board'],
      writesEnabled: false,
    });
  });
  try {
    const client = new AiChatClient({ baseUrl: stub.url });
    const status = await client.status();
    assert.equal(seen.method, 'GET');
    assert.equal(seen.url, '/api/v1/status');
    assert.equal(seen.authorization, undefined);
    assert.deepEqual(status, {
      version: '0.1.0',
      model: 'claude-opus-4-8',
      engine: 'postgresql',
      ready: true,
      features: ['chat', 'agent', 'board'],
      writesEnabled: false,
    });
  } finally {
    await stub.close();
  }
});

test('trailing slash on baseUrl does not double the path separator', async () => {
  let url;
  const stub = await startStub((req, res) => {
    url = req.url;
    sendJson(res, 200, []);
  });
  try {
    const client = new AiChatClient({ baseUrl: `${stub.url}/`, token: 't' });
    await client.listConversations();
    assert.equal(url, '/api/v1/conversations');
  } finally {
    await stub.close();
  }
});

test('conversations CRUD: list, get, delete', async () => {
  const conversation = {
    id: 'conv-9',
    title: 'Churned customers in June',
    createdAt: '2026-07-01T10:00:00Z',
    updatedAt: '2026-07-01T10:05:00Z',
  };
  const requests = [];
  const stub = await startStub((req, res) => {
    requests.push({ method: req.method, url: req.url, authorization: req.headers.authorization });
    if (req.method === 'GET' && req.url === '/api/v1/conversations') {
      sendJson(res, 200, [conversation]);
    } else if (req.method === 'GET' && req.url === '/api/v1/conversations/conv-9') {
      sendJson(res, 200, {
        conversation,
        messages: [
          { id: 'm1', role: 'user', text: 'Which customers churned in June?', createdAt: '2026-07-01T10:00:00Z' },
          {
            id: 'm2',
            role: 'assistant',
            text: 'Three customers churned.',
            sql: 'SELECT ...',
            result: { columns: ['name'], rowSample: [['acme']], rowCount: 3, truncated: false, executionTimeMs: 12 },
            createdAt: '2026-07-01T10:00:05Z',
          },
        ],
      });
    } else if (req.method === 'DELETE' && req.url === '/api/v1/conversations/conv-9') {
      sendJson(res, 200, { deleted: true });
    } else {
      sendJson(res, 404, { error: 'unknown resource', code: 'NOT_FOUND' });
    }
  });
  try {
    const client = new AiChatClient({ baseUrl: stub.url, token: 'tok' });

    const list = await client.listConversations();
    assert.deepEqual(list, [conversation]);

    const detail = await client.getConversation('conv-9');
    assert.deepEqual(detail.conversation, conversation);
    assert.equal(detail.messages.length, 2);
    assert.equal(detail.messages[0].role, 'user');
    assert.equal(detail.messages[1].sql, 'SELECT ...');
    assert.equal(detail.messages[1].result.rowCount, 3);

    const deleted = await client.deleteConversation('conv-9');
    assert.equal(deleted, undefined);

    assert.deepEqual(
      requests.map((r) => `${r.method} ${r.url}`),
      [
        'GET /api/v1/conversations',
        'GET /api/v1/conversations/conv-9',
        'DELETE /api/v1/conversations/conv-9',
      ],
    );
    assert.ok(requests.every((r) => r.authorization === 'Bearer tok'));
  } finally {
    await stub.close();
  }
});

test('getConversation() URL-encodes the id and maps 404 to NOT_FOUND', async () => {
  let url;
  const stub = await startStub((req, res) => {
    url = req.url;
    sendJson(res, 404, { error: 'unknown conversation', code: 'NOT_FOUND' });
  });
  try {
    const client = new AiChatClient({ baseUrl: stub.url, token: 't' });
    await assert.rejects(client.getConversation('a/b c'), (err) => {
      assert.ok(err instanceof AiChatError);
      assert.equal(err.status, 404);
      assert.equal(err.code, 'NOT_FOUND');
      return true;
    });
    assert.equal(url, '/api/v1/conversations/a%2Fb%20c');
  } finally {
    await stub.close();
  }
});

test('token admin endpoints: list, create, revoke, and ADMIN_REQUIRED mapping', async () => {
  const info = {
    id: 'tok-1',
    label: 'reporting-cron',
    fingerprint: 'ab12cd34',
    createdAt: '2026-07-01T09:00:00Z',
  };
  const requests = [];
  const stub = await startStub(async (req, res) => {
    const entry = { method: req.method, url: req.url };
    if (req.method === 'GET' && req.url === '/api/v1/tokens') {
      sendJson(res, 200, [{ ...info, lastUsedAt: '2026-07-02T09:00:00Z' }]);
    } else if (req.method === 'POST' && req.url === '/api/v1/tokens') {
      entry.body = JSON.parse(await readBody(req));
      sendJson(res, 201, { ...info, token: 'raw-secret-token' });
    } else if (req.method === 'DELETE' && req.url === '/api/v1/tokens/tok-1') {
      sendJson(res, 200, { deleted: true });
    } else {
      sendJson(res, 404, { error: 'unknown resource', code: 'NOT_FOUND' });
    }
    requests.push(entry);
  });
  try {
    const client = new AiChatClient({ baseUrl: stub.url, token: 'primary' });

    const tokens = await client.listTokens();
    assert.equal(tokens.length, 1);
    assert.equal(tokens[0].fingerprint, 'ab12cd34');
    assert.equal(tokens[0].lastUsedAt, '2026-07-02T09:00:00Z');

    const created = await client.createToken('reporting-cron');
    assert.equal(created.token, 'raw-secret-token');
    assert.equal(created.label, 'reporting-cron');
    assert.deepEqual(requests[1].body, { label: 'reporting-cron' });

    assert.equal(await client.revokeToken('tok-1'), undefined);
    assert.deepEqual(
      requests.map((r) => `${r.method} ${r.url}`),
      ['GET /api/v1/tokens', 'POST /api/v1/tokens', 'DELETE /api/v1/tokens/tok-1'],
    );
  } finally {
    await stub.close();
  }

  const adminStub = await startStub((req, res) => {
    sendJson(res, 403, { error: 'named tokens cannot manage tokens', code: 'ADMIN_REQUIRED' });
  });
  try {
    const named = new AiChatClient({ baseUrl: adminStub.url, token: 'named-token' });
    await assert.rejects(named.createToken('nope'), (err) => {
      assert.ok(err instanceof AiChatError);
      assert.equal(err.status, 403);
      assert.equal(err.code, 'ADMIN_REQUIRED');
      return true;
    });
  } finally {
    await adminStub.close();
  }
});
